### ISSUE TITLE
iOS and Voiceover ignores the focus trap trick for modal dialogs

### ISSUE THREAD

**Issue by julesongithub** [twbs/bootstrap#26770]
In order to avoid content, below the modal, being reading by AT, focus is kept within the modal when ever an attempt is made to tab out of the modal.

Unfortunately, this doesn't work whilst swiping in iOS with Voiceover (I'm not sure about Android but that's probably the same).

What's needed is something to aria-hide everything, except the modal, whilst it's open:

```js
var backgroundAriaHider = (function() {
    var $hide, $wrap;
            return {
                hideAllExcept: function (elementOrSelector) {
                    var $node = elementOrSelector instanceof jQuery ? elementOrSelector : $(elementOrSelector);
                    $hide = $();
                    $wrap = $();
                    do {
                        $hide = $hide.add($node.siblings(':not(script):not([aria-hidden="true"])'));
                        // siblings doesn't include plain text elements; wrap these in a span
                        $node.parent().contents().each(function () {
                            if (this.nodeType === 3 && this.nodeValue.trim().length) {
                                $wrap = $wrap.add(this);
                            }
                        });
                        $node = $node.parent();
                    } while ($node[0].tagName.toLowerCase() !== 'body');
                    $hide.attr('aria-hidden', true);
                    $wrap.wrap('<span aria-hidden="true" />');
                },
                reset: function () {
                    $hide.removeAttr('aria-hidden');
                    $wrap.unwrap();
                    $hide = null;
                    $wrap = null;
                }
            };
})();
```
Once the modal is open, and has been focussed, call  `backgroundAriaHider.hideAllExcept($modal);`
Then once, it's closing call `backgroundAriaHider.reset();`

This is just a quick example; for instance you might want to take note of where an element has an explicit [aria-hidden="false"] and return it to that state instead of removing the attribute.

Cheers,
Jules

**Comment by patrickhlauke** [comment]
i had started reworking modals ages ago using the `inert` attribute https://github.com/patrickhlauke/bootstrap/tree/v4-dev-modal-rework-1 ... but work on that stalled as it still needs a polyfill (and a slight rework of what bootstrap should/shouldn't allow for modals...they always need to be in a separate DOM branch from the rest of the page that is made `inert` - or `aria-hidden`, while currently it's allowed to keep modals in place in the DOM).

but yes, agree this needs tackling. just wondering if we could revisit the approach using `inert` + polyfill, rather than custom "add `aria-hidden`, and neutralize/make unfocusable any other focusable element" (which would allow us to remove the hidden focus trap/handler). @mdo ?

**Comment by julesongithub** [comment]
I'd never heard of the inert tag and just tried the Google and WICG polyfills on my phone. The Google one doesn't work but the WICG one does because it also adds the [aria-hidden] attribute along with [inert].

Your technique would work by simply using the [aria-hidden] attribute instead of the [inert] tag but I just wonder if there's situations where the dialog needs to remain in place and not detached and appended outside of the inert section.

**Comment by patrickhlauke** [comment]
just adding `aria-hidden` and keeping the current focus trap does work, but conceptually it's a lot cleaner to implement `inert` (if it were supported) as that not only hides the content from the accessibility tree like `aria-hidden` does, but also makes any other focusable things in the non-focusable (even without assitive tech running), avoiding the need for custom focus handling acrobatics.

as an interim solution, however, may scale down and just do the `aria-hidden` on the main page part for now and keep the focus trap. once i get time...

**Comment by patrickhlauke** [comment]
>  I just wonder if there's situations where the dialog needs to remain in place and not detached and appended outside of the inert section.

i think historically there's been some uses of dialogs inside only portions of a page, or tricky z-index things that required the modal to not be hanging straight off the `<body>` but somewhere else. i may investigate a bit further, time permitting

**Comment by arnaldomm** [comment]
The dialog box should have these two attributes:
role="dialog"
aria-modal="true"

Be sure to test on good screen readers-browser combinations:
- VoiceOver: Safari.
- JAWS: Chrome, Firefox, Internet Explorer.
- NVDA: Firefox, Chrome.
- Narrator: Microsoft Edge

**Comment by camtin** [comment]
I can confirm at least that aria-modal="true" and role="dialog" work on:

macOS v10.14 Mojave VoiceOver and Safari v12
macOS v10.14 Mojave VoiceOver and Chrome v88.0.4324.192

The VoiceOver virtual cursor ignores all content 'behind' the modal.

This does not work on:
macOS v10.14 Mojave VoiceOver and Firefox v86.0.1
